### PR TITLE
fix(integrations): restore strict Qwen3 bitwise parity

### LINKED ISSUE #335: [Vime][TP=2 CP=2] Strict R/R linear_logp remains nonzero (0.023694925) after shared backend
## Summary

We ran Vime R/R (rollout/reference versus training/reference) on 8x H100 with Qwen3-8B, CP=2, TP=2. Both rollout and training report the same deterministic RL-Kernel operator providers, including the separate strict `linear_logp` backend:

```text
attention:  rlkernel.attention.deterministic.v1
ffn:        rlkernel.ffn.qwen3.deterministic.v1
linear_logp: rlkernel.linear_logp.bitwise.v1
```

The end-to-end job completed rollout and training and wrote an Nsight trace, but failed Vime's zero-threshold assertion:

```text
train/train_rollout_logprob_abs_diff = 0.023694925010204315
threshold = 0
```

This is **not** a strict-bitwise pass. The metric above is Vime's `sum_of_sample_mean(abs(...))`; it is not the maximum per-element absolute difference.

The immediate goal is to locate and eliminate the remaining difference source so R/R satisfies elementwise `mismatch_count == 0`, `max_abs_diff == 0`, and the Vime aggregate is exactly zero.

## Environment and Reproduction

| Item | Value |
| --- | --- |
| Host | `xe9680-cn-2` |
| GPU | 8x H100 |
| RL-Kernel checkout | `/home/ellm/ljj/RL-Kernel-pr230-final` |
| RL-Kernel branch/base | `codex/pr230-h100-final` / `2717867` |
| Vime checkout | `/home/ellm/ljj/vime` |
| Ray job | `raysubmit_EaaERakEkuwF1G5S` |
| Run directory | `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict` |
| Launcher | `/home/ellm/ljj/vime/scripts/run-qwen3-8B-rlkernel-tp2-cp2-trace.sh` |
| Nsight trace | `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/trace/qwen3-8b-tp2-cp2.nsys-rep` |
| Ray driver log | `/tmp/ray/session_latest/logs/job-driver-raysubmit_EaaERakEkuwF1G5S.log` |

The strict-R/R environment is configured in `/home/ellm/ljj/vime/scripts/run-qwen3-8B-rlkernel-tp2-cp2-trace.sh` at lines 36-49. The strict provider selection and zero threshold are at lines 191-202. Nsight configuration is at lines 222-237.

### 1. The strict linear_logp route was selected on both sides

The wrapper is deliberately an independent strict backend, not a replacement for the original performance-oriented `linear_logp` path:

- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/integrations/linear_logp.py:95-101`: independent strict-backend definition.
- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/integrations/linear_logp.py:134-249`: contract validation.
- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/integrations/linear_logp.py:251-277`: mismatch-axis metadata.
- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/integrations/linear_logp.py:279-361`: strict operator call.

The CUDA TP implementation is reached through:

- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/kernels/ops/cuda/loss/linear_logp.py:1058-1128`
- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/kernels/ops/cuda/loss/linear_logp.py:1198-1219`

Runtime strict-logp contract:

```text
real vocab:   151936
padded vocab: 152064
TP=2 shards:  [76032, 4096]
```

All wrapper mismatch axes were reported `false`:

```text
vocab_shard_ownership: false
selected_token_identity: false
vocab_lse_reduction: false
```

### 2. Rollout captures the actual vLLM hidden states and LM-head weight, then invokes strict linear_logp

- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/integrations/vllm_runtime.py:190-225`: vLLM hidden-state and LM-head capture.
- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/integrations/framework_operators.py:1195-1243`: vLLM sampler score replacement.
- `/home/ellm/ljj/RL-Kernel-pr230-final/rl_engine/integrations/megatron_runtime.py:99-108`: Megatron integration.

Rollout readback:

`/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/readbacks/vllm-rollout-3551824.json`

Observed call counts:

```text
attention/ffn calls: 37008
strict linear_logp calls: 1026
```

### 3. Native vLLM score and strict recomputed score differ

The rollout readback's diagnostic compares native vLLM token scores against the strict recomputed values. It has 13 differing rows out of 15 and a maximum absolute difference of `0.13068890571594238`.

This is a diagnostic comparison, **not** the failed train-versus-rollout assertion. It confirms that the rollout still obtains token selection from the native vLLM sampler before score replacement, and that the two score routes cannot be assumed numerically identical.

### 4. Training-side readbacks exist on all nonzero ranks

- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/readbacks/megatron-training-3549956.json`
- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/readbacks/megatron-training-3550476.json`
- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/readbacks/megatron-training-3550478.json`
- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/readbacks/megatron-training-3550479.json`

## Missing Diagnostics: Rank 0 Is Lost Before Data Is Saved

The first failure prevents rank 0 from writing the debugging artifact required to compare the values. The exact control flow is:

1. The comparison is calculated in `/home/ellm/ljj/vime/vime/backends/megatron_utils/loss.py:1125-1140`.
2. The zero-threshold assertion is raised in `/home/ellm/ljj/vime/vime/backends/megatron_utils/model.py:998-999`.
3. `save_debug_train_data()` runs only after `train()` returns, in `/home/ellm/ljj/vime/vime/backends/megatron_utils/actor.py:520`.
4. The writer is `/home/ellm/ljj/vime/vime/utils/train_dump_utils.py:10-28`.

The result is that these files exist:

- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/train-data/0.rank1.pt`
- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/train-data/0.rank2.pt`
- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/train-data/0.rank3.pt`

But the decisive rank-0 file is missing:

- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/train-data/0.rank0.pt`

The aggregate validator report is missing for the same reason:

- `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/bitwise-validation.json`

The post-job validator invocation itself is only reached after a successful Ray job in `/home/ellm/ljj/vime/scripts/run-qwen3-8B-rlkernel-tp2-cp2-trace.sh:276-279`.

## Candidate Difference Sources (Not Yet Proven)

The evidence does not yet identify the source. The following are hypotheses to test, not conclusions:

1. **Pre-LM-head hidden-state divergence.** vLLM decode/paged-KV execution and Megatron packed/CP execution may produce different hidden states before the shared strict `linear_logp` call, despite using the same named attention and FFN backends.
2. **Row/token identity or ordering mismatch.** Vime currently flattens, concatenates, and subtracts values without validating a stable row key or a permutation. The intended structural request construction is in `/home/ellm/ljj/vime/vime/backends/megatron_utils/loss.py:540-609`.
3. **Native vLLM sampler semantics leak into the score contract.** Token selection is native first and score replacement happens after. The native-versus-strict diagnostic shows those routes are different.
4. **Temperature/logit semantics differ.** The placement, dtype, or interpretation of temperature and already-scaled logits may not be identical even when the configured temperature is `1.0`.
5. **Insufficient failure-path artifacts.** Without rank-0 values and first-mismatch metadata, it is impossible to distinguish a true arithmetic difference from a comparison alignment problem.

## Requested Instrumentation and Acceptance Criteria

Before the next full R/R run, please make the failure path persist diagnostics before raising:

1. Validate stable per-row keys and selected token IDs before comparing rollout and training logprobs; fail separately on cardinality, key, token-ID, or ordering mismatch.
2. For the first mismatching element, persist the row key, token ID, hidden-state hash, LM-head checksum, temperature, local selected value, local LSE, merged LSE, and final FP32 logp on both rollout and training sides.
3. Persist `0.rank0.pt` and `bitwise-validation.json` before the zero-threshold assertion, including elementwise `mismatch_count` and `max_abs_diff`.
4. Keep `rlkernel.linear_logp.bitwise.v1` separate from the original high-performance `linear_logp` implementation.
5. Re-run the same TP=2, CP=2 Vime R/R configuration and require all of the following:

```text
mismatch_count == 0
max_abs_diff == 0
train/train_rollout_logprob_abs_diff == 0
```

Once the arithmetic comparison passes, inspect `/home/ellm/ljj/vime/outputs/rlkernel/rr-20260823-235754-strict/trace/qwen3-8b-tp2-cp2.nsys-rep` and file a separate local performance issue identifying the dominant kernels and their cost.

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Summary

- add the strict FA4 BSHD entrypoint plus bundled-vLLM FA4/rotary compatibility
- align Megatron and vLLM Attention projections on DetGemmOp and match Megatron's RMSNorm arithmetic
- replace every duplicate sampled-token logprob column before vLLM response conversion

## Validation

- `python -m pytest -q tests/test_framework_runtime_adapters.py tests/test_flashinfer_pr7_attention.py`: 77 passed, 1 skipped
- 8x H100, Qwen3-8B Dense, TP=2, CP=2, 128 response tokens, debug environment variables unset:
  - `train_rollout_logprob_abs_diff: 0.0`
  - `mismatch_count: 0 / 128`
  - `max_abs_diff: 0.0`

The validation job subsequently hits Vime/Megatron's existing distributed optimizer checkpoint assertion `(torch.Size([]), 0, 311427072)`, after the strict consistency metric has passed.

Fixes #335

### CODE REVIEW

**Review by Flink-ddd — APPROVED** [review]
LGTM, Thank you for update.
